// @curvefi/api/lib/constants/coins/ethereum.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.aTokensEthereum = exports.ycTokensEthereum = exports.yTokensEthereum = exports.cTokensEthereum = exports.COINS_ETHEREUM = exports.USD_COINS_ETHEREUM = exports.EUR_COINS_ETHEREUM = exports.LINK_COINS_ETHEREUM = exports.ETH_COINS_ETHEREUM = exports.BTC_COINS_ETHEREUM = void 0;
var utils_1 = require("../utils");
exports.BTC_COINS_ETHEREUM = {
    sbtccrv: "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3",
    hbtc: "0x0316EB71485b0Ab14103307bf65a021042c6d380",
    renbtc: "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D",
    wbtc: "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    tbtc: "0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa",
    pbtc: "0x5228a22e72ccC52d415EcFd199F99D0665E7733b",
    bbtc: "0x9be89d2a4cd102d8fecc6bf9da793be995c22541",
    obtc: "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68",
    sbtc: "0xfE18be6b3Bd88A2D2A7f928d00292E7a9963CfC6", // sBTC
};
exports.ETH_COINS_ETHEREUM = {
    steth: "0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84",
    eth: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
    ankreth: "0xE95A203B1a91a908F9B9CE46459d101078c2c3cb",
    seth: "0x5e74c9036fb86bd7ecdcb084a0673efc32ea31cb",
    reth: "0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593",
    weth: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    frxeth: "0x5E8422345238F34275888049021821E8E08CAa1f",
    xaut: "0x68749665ff8d2d112fa859aa293f07a622782f38", // XAUt TODO move to GOLD
};
exports.LINK_COINS_ETHEREUM = {
    link: "0x514910771AF9Ca656af840dff83E8264EcF986CA",
    slink: "0xbBC455cb4F1B9e4bFC4B73970d360c8f032EfEE6", // sLINK
};
exports.EUR_COINS_ETHEREUM = {
    eurs: "0xdB25f211AB05b1c97D595516F45794528a807ad8",
    seur: "0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
    eurt: "0xC581b735A1688071A1746c968e0798D642EDE491",
    euroc: "0x1aBaEA1f7C830bD89Acc67eC4af516284b1bC33c",
};
exports.USD_COINS_ETHEREUM = {
    ycdai: "0x99d1Fa417f94dcD62BfE781a1213c092a47041Bc",
    ycusdc: "0x9777d7E2b60bB01759D0E2f8be2095df444cb07E",
    ycusdt: "0x1bE5d71F2dA660BFdee8012dDc58D024448A0A59",
    pax: "0x8E870D67F660D95d5be530380D0eC0bd388289E1",
    adai: "0x028171bCA77440897B824Ca71D1c56caC55b68A3",
    ausdc: "0xBcca60bB61934080951369a648Fb03DF4F96263C",
    ausdt: "0x3Ed3B47Dd13EC9a98b44e6204A523E766B225811",
    asusd: "0x6c5024cd4f8a59110119c56f8933403a539555eb",
    cdai: "0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643",
    cusdc: "0x39AA39c021dfbaE8faC545936693aC917d5E7563",
    cydai: "0x8e595470ed749b85c6f7669de83eae304c2ec68f",
    cyusdc: "0x76eb2fe28b36b3ee97f3adae0c69606eedb2a37c",
    cyusdt: "0x48759f220ed983db51fa7a8c0d2aab8f3ce4166a",
    bydai: "0xC2cB1040220768554cf699b0d863A3cd4324ce32",
    byusdc: "0x26EA744E5B887E5205727f55dFBE8685e3b21951",
    byusdt: "0xE6354ed5bC4b393a5Aad09f21c46E101e692d447",
    ybusd: "0x04bC0Ab673d88aE9dbC9DA2380cB6B79C4BCa9aE",
    ydai: "0x16de59092dAE5CcF4A1E6439D611fd0653f0Bd01",
    yusdc: "0xd6aD7a6750A7593E092a9B218d66C0A814a3436e",
    yusdt: "0x83f798e925BcD4017Eb265844FDDAbb448f1707D",
    ytusd: "0x73a052500105205d34Daf004eAb301916DA8190f",
    gusd: "0x056Fd409E1d7A124BD7017459dFEa2F387b6d5Cd",
    husd: "0xdF574c24545E5FfEcb9a659c229253D4111d87e1",
    usdk: "0x1c48f86ae57291F7686349F12601910BD8D470bb",
    musd: "0xe2f2a5C287993345a840Db3B0845fbC70f5935a5",
    rsv: "0x196f4727526eA7FB1e17b2071B3d8eAA38486988",
    dusd: "0x5BC25f649fc4e26069dDF4cF4010F9f706c23831",
    ust: "0xa47c8bf37f92abed4a126bda807a7b7498661acd",
    usdp: "0x1456688345527bE1f37E9e627DA0837D6f08C925",
    usdn: "0x674C6Ad92Fd080e4004b2312b45f796a192D27a0",
    dai: "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    usdc: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    usdt: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
    susd: "0x57Ab1ec28D129707052df4dF418D58a2D46d5f51",
    tusd: "0x0000000000085d4780B73119b644AE5ecd22b376",
    frax: "0x853d955acef822db058eb8505911ed77f175b99e",
    lusd: "0x5f98805A4E8be255a32880FDeC7F6728C6568bA0",
    busd: "0x4Fabb145d64652a948d72533023f6E7A623C7C53",
    alusd: "0xbc6da0fe9ad5f3b0d58160288917aa56653660e9",
    mim: "0x99d8a9c45b2eca8864373a26d1459e3dff1e17f3",
    rai: "0x03ab458634910aad20ef5f1c8ee96f1d6ac54919",
    wormholeust: "0xa693B19d2931d498c5B318dF961919BB4aee87a5",
    '3crv': "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490",
    crv: "0xD533a949740bb3306d119CC777fa900bA034cd52",
    cvx: "0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b", // CVX
};
exports.COINS_ETHEREUM = (0, utils_1.lowerCaseValues)(__assign(__assign(__assign(__assign(__assign(__assign({}, exports.BTC_COINS_ETHEREUM), exports.ETH_COINS_ETHEREUM), exports.LINK_COINS_ETHEREUM), exports.EUR_COINS_ETHEREUM), exports.USD_COINS_ETHEREUM), { snx: "0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f", spell: "0x090185f2135308bad17527004364ebcc2d37e5f6", t: "0xCdF7028ceAB81fA0C6971208e83fa7872994beE5" }));
var DECIMALS_ETHEREUM = {
    "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3": 18,
    "0x0316EB71485b0Ab14103307bf65a021042c6d380": 18,
    "0x99d1Fa417f94dcD62BfE781a1213c092a47041Bc": 18,
    "0x9777d7E2b60bB01759D0E2f8be2095df444cb07E": 6,
    "0x1bE5d71F2dA660BFdee8012dDc58D024448A0A59": 6,
    "0x8E870D67F660D95d5be530380D0eC0bd388289E1": 18,
    "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D": 8,
    "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599": 8,
    "0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa": 18,
    "0x5228a22e72ccC52d415EcFd199F99D0665E7733b": 18,
    "0x9be89d2a4cd102d8fecc6bf9da793be995c22541": 8,
    "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68": 18,
    "0xfE18be6b3Bd88A2D2A7f928d00292E7a9963CfC6": 18,
    "0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84": 18,
    "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE": 18,
    "0xE95A203B1a91a908F9B9CE46459d101078c2c3cb": 18,
    "0x5e74c9036fb86bd7ecdcb084a0673efc32ea31cb": 18,
    "0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593": 18,
    "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2": 18,
    "0xdB25f211AB05b1c97D595516F45794528a807ad8": 2,
    "0xD71eCFF9342A5Ced620049e616c5035F1dB98620": 18,
    "0xC581b735A1688071A1746c968e0798D642EDE491": 6,
    "0x514910771AF9Ca656af840dff83E8264EcF986CA": 18,
    "0xbBC455cb4F1B9e4bFC4B73970d360c8f032EfEE6": 18,
    "0x028171bCA77440897B824Ca71D1c56caC55b68A3": 18,
    "0xBcca60bB61934080951369a648Fb03DF4F96263C": 6,
    "0x3Ed3B47Dd13EC9a98b44e6204A523E766B225811": 6,
    "0x6c5024cd4f8a59110119c56f8933403a539555eb": 18,
    "0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643": 8,
    "0x39AA39c021dfbaE8faC545936693aC917d5E7563": 8,
    "0x8e595470ed749b85c6f7669de83eae304c2ec68f": 8,
    "0x76eb2fe28b36b3ee97f3adae0c69606eedb2a37c": 8,
    "0x48759f220ed983db51fa7a8c0d2aab8f3ce4166a": 8,
    "0xC2cB1040220768554cf699b0d863A3cd4324ce32": 18,
    "0x26EA744E5B887E5205727f55dFBE8685e3b21951": 6,
    "0xE6354ed5bC4b393a5Aad09f21c46E101e692d447": 6,
    "0x04bC0Ab673d88aE9dbC9DA2380cB6B79C4BCa9aE": 18,
    "0x16de59092dAE5CcF4A1E6439D611fd0653f0Bd01": 18,
    "0xd6aD7a6750A7593E092a9B218d66C0A814a3436e": 6,
    "0x83f798e925BcD4017Eb265844FDDAbb448f1707D": 6,
    "0x73a052500105205d34Daf004eAb301916DA8190f": 18,
    "0x056Fd409E1d7A124BD7017459dFEa2F387b6d5Cd": 2,
    "0xdF574c24545E5FfEcb9a659c229253D4111d87e1": 8,
    "0x1c48f86ae57291F7686349F12601910BD8D470bb": 18,
    "0xe2f2a5C287993345a840Db3B0845fbC70f5935a5": 18,
    "0x196f4727526eA7FB1e17b2071B3d8eAA38486988": 18,
    "0x5BC25f649fc4e26069dDF4cF4010F9f706c23831": 18,
    "0xa47c8bf37f92abed4a126bda807a7b7498661acd": 18,
    "0x1456688345527bE1f37E9e627DA0837D6f08C925": 18,
    "0x674C6Ad92Fd080e4004b2312b45f796a192D27a0": 18,
    "0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f": 18,
    "0x6B175474E89094C44Da98b954EedeAC495271d0F": 18,
    "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48": 6,
    "0xdAC17F958D2ee523a2206206994597C13D831ec7": 6,
    "0x57Ab1ec28D129707052df4dF418D58a2D46d5f51": 18,
    "0x0000000000085d4780B73119b644AE5ecd22b376": 18,
    "0x853d955acef822db058eb8505911ed77f175b99e": 18,
    "0x5f98805A4E8be255a32880FDeC7F6728C6568bA0": 18,
    "0x4Fabb145d64652a948d72533023f6E7A623C7C53": 18,
    "0xbc6da0fe9ad5f3b0d58160288917aa56653660e9": 18,
    "0x99d8a9c45b2eca8864373a26d1459e3dff1e17f3": 18,
    "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490": 18,
    "0xD533a949740bb3306d119CC777fa900bA034cd52": 18,
    "0x4e3fbd56cd56c3e72c1403e103b45db9da5b9d2b": 18,
    "0x68749665ff8d2d112fa859aa293f07a622782f38": 6,
    "0x090185f2135308bad17527004364ebcc2d37e5f6": 18,
    "0xCdF7028ceAB81fA0C6971208e83fa7872994beE5": 18,
    "0xa693B19d2931d498c5B318dF961919BB4aee87a5": 6,
    "0x03ab458634910aad20ef5f1c8ee96f1d6ac54919": 18,
    // --- REWARD TOKENS ---
    "0xa3BeD4E1c75D00fa6f4E5E6922DB7261B5E9AcD2": 18,
    "0x8762db106B2c2A0bccB3A80d1Ed41273552616E8": 18,
    "0x20c36f062a31865bED8a5B1e512D9a1A20AA333A": 18,
    "0x89Ab32156e46F46D02ade3FEcbe5Fc4243B9AAeD": 18,
    "0x3c9d6c1C73b31c837832c72E04D3152f051fc1A9": 18,
    "0x4da27a545c0c5b758a6ba100e3a049001de870f5": 18,
    "0x5a98fcbea516cf06857215779fd812ca3bef1b32": 18,
    "0xe0ad1806fd3e7edf6ff52fdb822432e847411033": 18,
    "0x8290333cef9e6d528dd5618fb97a76f268f3edd4": 18,
    "0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0": 18,
    "0x6dea81c8171d0ba574754ef6f8b412f2ed88c54d": 18,
    "0xef3a930e1ffffacd2fc13434ac81bd278b0ecc8d": 18,
    "0xdbdb4d16eda451d0503b854cf79d55697f90c8df": 18, // ALCX
};
exports.cTokensEthereum = [
    '0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643',
    '0x39AA39c021dfbaE8faC545936693aC917d5E7563',
    "0x8e595470ed749b85c6f7669de83eae304c2ec68f",
    "0x48759f220ed983db51fa7a8c0d2aab8f3ce4166a",
    "0x76eb2fe28b36b3ee97f3adae0c69606eedb2a37c", // cyUSDC
].map(function (a) { return a.toLowerCase(); });
exports.yTokensEthereum = [
    "0xC2cB1040220768554cf699b0d863A3cd4324ce32",
    "0x26EA744E5B887E5205727f55dFBE8685e3b21951",
    "0xE6354ed5bC4b393a5Aad09f21c46E101e692d447",
    "0x16de59092dAE5CcF4A1E6439D611fd0653f0Bd01",
    "0xd6aD7a6750A7593E092a9B218d66C0A814a3436e",
    "0x83f798e925BcD4017Eb265844FDDAbb448f1707D",
    "0x04bC0Ab673d88aE9dbC9DA2380cB6B79C4BCa9aE",
    "0x73a052500105205d34Daf004eAb301916DA8190f", // yTUSD
].map(function (a) { return a.toLowerCase(); });
exports.ycTokensEthereum = [
    "0x99d1Fa417f94dcD62BfE781a1213c092a47041Bc",
    "0x9777d7E2b60bB01759D0E2f8be2095df444cb07E",
    "0x1bE5d71F2dA660BFdee8012dDc58D024448A0A59", // ycUSDT
].map(function (a) { return a.toLowerCase(); });
exports.aTokensEthereum = [
    "0x028171bCA77440897B824Ca71D1c56caC55b68A3",
    "0xBcca60bB61934080951369a648Fb03DF4F96263C",
    "0x3Ed3B47Dd13EC9a98b44e6204A523E766B225811",
    "0x6c5024cd4f8a59110119c56f8933403a539555eb", // sSUSD
].map(function (a) { return a.toLowerCase(); });
